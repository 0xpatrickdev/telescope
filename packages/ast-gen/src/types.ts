export declare type fieldType = 'Long' | 'Coin' | 'Duration' | 'Height' | string;

export interface Field {
    name: string;
    type: fieldType;
    node: any;
};
export interface Interface {
    name: string;
    fields: Field[]
};
export interface MessageSchema extends Interface {
    typeUrl: string;
    name: string;
    fields: Field[]
};

export interface EnumConverter {
    funcName: string;
    convertType: 'to' | 'from';
    type: string;
};
export interface Enum {
    filename?: string;

    name: string;
    node: any; // ast
    from?: EnumConverter;
    to?: EnumConverter;
}
export interface Mutation {
    typeUrl: string;
    TypeName: string;
    methodName: string;
}

export interface AminoException {
    aminoType?: string;
    toAmino?: object;
    fromAmino?: object;
    AminoMsg?: object;
    Msg?: object;
}
export interface AminoExceptions {
    [key: string]: AminoException;
}

export const DEFAULT_AMINO_EXCEPTIONS: AminoExceptions = {
    '/cosmos.distribution.v1beta1.MsgSetWithdrawAddress': {
        aminoType: 'cosmos-sdk/MsgModifyWithdrawAddress'
    },
    '/cosmos.distribution.v1beta1.MsgWithdrawDelegatorReward': {
        aminoType: 'cosmos-sdk/MsgWithdrawDelegationReward'
    }
}