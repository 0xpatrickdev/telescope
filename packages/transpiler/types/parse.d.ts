import { ProtoStore, ProtoRef } from '@osmonauts/proto-parser';
import { ProtoField } from '@osmonauts/ast';
export declare const parse: (store: ProtoStore, ref: ProtoRef) => any;
export declare const getParsedObjectName: (ref: ProtoRef, obj: any, scope: string[]) => string;
export declare const getKeyTypeObjectName: (obj: any, field: ProtoField) => string;
export declare const parseType: (store: ProtoStore, ref: ProtoRef, obj: any, imports: object, body: any[], scope: string[], isNested?: boolean) => void;
export declare const parseEnum: (store: ProtoStore, ref: ProtoRef, obj: any, imports: object, body: any[], scope: string[], isNested?: boolean) => void;
interface ParseRecur {
    store: ProtoStore;
    ref: ProtoRef;
    obj: any;
    imports: object;
    body: any[];
    scope: string[];
    isNested: boolean;
}
export declare const parseRecur: ({ store, ref, obj, imports, body, scope, isNested }: ParseRecur) => void;
export {};
