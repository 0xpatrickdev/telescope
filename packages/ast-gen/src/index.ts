export * from './types';
export * from './amino';
export * from './client';
export * from './messages';
export * from './utils';