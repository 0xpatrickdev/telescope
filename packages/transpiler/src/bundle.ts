import { ProtoStore } from '@osmonauts/proto-parser';
import { recursiveModuleBundle, importNamespace } from '@osmonauts/ast';
import * as dotty from 'dotty';
import { TelescopeInput } from './index';
import { join, relative, dirname } from 'path';

// TODO move to store
export const getPackages = (store: ProtoStore) => {
    return store.getProtos().reduce((m, proto) => {
        if (proto.proto.package) {
            m[proto.proto.package] = m[proto.proto.package] || []
            m[proto.proto.package].push(proto);
        }
        return m;
    }, {});
}

export const getPackagesBundled = (store: ProtoStore) => {
    const objectified = {};
    const pkgs = getPackages(store);
    Object.keys(pkgs).forEach(key => {
        const files = pkgs[key];
        dotty.put(objectified, key, {
            pkg: key,
            files
        });
    });
    return objectified;

}

export const bundlePackages = (store: ProtoStore, input: TelescopeInput) => {
    const allPackages = getPackagesBundled(store);
    return Object.keys(allPackages).map(base => {
        const pkgs = allPackages[base];
        const bundleVariables = {};
        const bundleFile = join(input.protoDir, base, 'bundle.ts');
        const importPaths = [];
        parsePackage(pkgs, bundleFile, importPaths, bundleVariables);
        const body = recursiveModuleBundle(bundleVariables);
        return {
            bundleFile,
            importPaths,
            base,
            body
        };
    });
}

export const parsePackage = (obj, bundleFile, importPaths, bundleVariables) => {
    if (!obj) return;
    if (obj.pkg && obj.files) {
        obj.files.forEach(file => {
            createFileBundle(obj, file.filename, bundleFile, importPaths, bundleVariables);
            // TODO better namespacing for aminos/messages/encoding:
            // file.outFiles.forEach(outFile =>
            //     createFileBundle(obj, outFile.filename, bundleFile, importPaths, bundleVariables)
            // );
        });
        return;
    }
    Object.keys(obj).forEach(mini => {
        parsePackage(obj[mini], bundleFile, importPaths, bundleVariables);
    })
}


let counter = 0;
const createFileBundle = (obj, filename, bundleFile, importPaths, bundleVariables) => {
    let rel = relative(dirname(bundleFile), filename);
    if (!rel.startsWith('.')) rel = `./${rel}`;
    const variable = `_${counter++}`;
    importPaths.push(importNamespace(variable, rel));
    dotty.put(bundleVariables, obj.pkg + '.__export', true);
    dotty.put(bundleVariables, obj.pkg + '.' + variable, true);
}