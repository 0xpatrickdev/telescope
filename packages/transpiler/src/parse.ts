import { ProtoStore, ProtoRoot, ProtoRef, getObjectName } from '@osmonauts/proto-parser';
import {
    createProtoType,
    createCreateProtoType,
    createProtoEnum,
    createProtoObjectWithMethods,
    ProtoType,
    ProtoParseContext,
    AminoParseContext,
    makeAminoTypeInterface,
    aminoConverter,
    getKeyTypeEntryName,
    // client
    createClient,
    GenericParseContext,
    // registry 
    createTypeRegistry,
    createRegistryLoader,
    // methods (needs refactor)
    toObjectWithPartialMethods,
    toObjectWithEncodedMethods,
    toObjectWithJsonMethods,
    toObjectWithToJSONMethods,
    toObjectWithFromJSONMethods
} from '@osmonauts/ast';
import { camel, snake } from 'case';

const getRoot = (ref: ProtoRef): ProtoRoot => {
    if (ref.traversed) return ref.traversed;
    return ref.proto;
};

export interface TelescopeParseContext {
    proto: ProtoParseContext;
    amino: AminoParseContext;
    store: ProtoStore;
    ref: ProtoRef;
    parsedImports: Record<string, any>;
    body: any[];
    mutations: any[];
    queries: any[];
    types: any[];
}

export const parse = (
    context: TelescopeParseContext,
): void => {
    const root = getRoot(context.ref);
    parseRecur({
        context,
        obj: root.root,
        scope: [],
        isNested: false
    });
};

export const getParsedObjectName = (
    ref: ProtoRef,
    obj: any,
    scope: string[],
) => {
    const root = getRoot(ref);
    const allButPackage = scope.splice(root.package.split('.').length);
    // pull off "this" name
    allButPackage.pop();
    return getObjectName(obj.name, [root.package, ...allButPackage]);
};

// TODO potentially move this back to ast or proto bc the ast lib references MapEntries...
const makeKeyTypeObj = (ref: ProtoRef, field: any, scope: string[]) => {
    const root = getRoot(ref);
    const scoped = scope.splice(root.package.split('.').length);
    const adhocObj: ProtoType = {
        comment: undefined,
        fields: {
            key: {
                id: 1,
                type: field.keyType,
                scope: [...scoped],
                parsedType: {
                    name: field.keyType,
                    type: field.keyType
                },
                comment: undefined,
                options: undefined
            },
            value: {
                id: 2,
                type: field.parsedType.name,
                scope: [...scoped],
                parsedType: {
                    name: field.type,
                    type: field.parsedType.type
                },
                comment: undefined,
                options: undefined
            }
        }
    };
    return adhocObj;
}

export const parseType = (
    context: TelescopeParseContext,
    obj: any,
    scope: string[],
    isNested: boolean = false
) => {

    obj.keyTypes.forEach(field => {
        const keyTypeObject = makeKeyTypeObj(context.ref, field, scope);
        const name = getParsedObjectName(context.ref, {
            name: getKeyTypeEntryName(obj.name, field.parsedType.name)
        }, scope);
        context.body.push(createProtoType(name, keyTypeObject));
        context.body.push(createCreateProtoType(name, keyTypeObject));
        context.body.push(createProtoObjectWithMethods(context.proto, name, keyTypeObject));
    })

    // parse nested names
    let name = obj.name;
    if (isNested) {
        name = getParsedObjectName(context.ref, obj, scope);
    }

    // context for types that can be exported
    if (!isNested) {
        context.types.push(
            {
                [name]: obj
            }
        );
    }

    context.body.push(createProtoType(name, obj));
    context.body.push(createCreateProtoType(name, obj));
    context.body.push(createProtoObjectWithMethods(context.proto, name, obj));

    // render nested LAST
    if (obj.nested) {
        Object.keys(obj.nested).forEach(key => {
            // isNested = true;
            parseRecur({
                context,
                obj: obj.nested[key],
                scope: [...scope, key],
                isNested: true
            });
        });
    }

};

export const parseEnum = (
    context: TelescopeParseContext,
    obj: any,
    scope: string[],
    isNested: boolean = false
) => {
    let name = obj.name;
    // parse nested names
    if (isNested) {
        name = getParsedObjectName(context.ref, obj, scope);
    }
    context.body.push(createProtoEnum(name, obj));
};

export const parseService = (
    context: TelescopeParseContext,
    obj: any,
    scope: string[],
    isNested: boolean = false
) => {

    const methodHash: Record<string, {
        requestType: string;
        responseType: string;
        comment?: string;
    }> = obj.methods;

    if (!['Msg', 'Query'].includes(obj.name)) {
        throw new Error('unsupported Service type. Contact Maintainers.');
    }

    const isMutation = obj.name === 'Msg';

    // NOTE
    // in future this can get hoisted up and we could theoretically aggregate multiple 
    // proto files in the same package for same-named services, e.g. two files with `service Msg`

    const lookups = Object.entries(methodHash)
        .map(([key, value]) => {
            const lookup = context.store.get(context.ref, value.requestType);
            if (!lookup) {
                console.warn(`cannot find ${value.requestType}`);
                throw new Error('undefined symbol for service.');
            }
            const lookupResponse = context.store.get(context.ref, value.responseType);
            if (!lookupResponse) {
                console.warn(`cannot find ${value.requestType}`);
                throw new Error('undefined symbol for service.');
            }
            (isMutation ? context.mutations : context.queries).push({
                methodName: key,
                package: context.ref.proto.package,
                message: lookup.importedName,
                messageImport: lookup.import ?? context.ref.filename,
                response: lookupResponse.importedName,
                responseImport: lookupResponse.import ?? context.ref.filename,
                comment: value.comment
            })
            return lookup;
        });


    if (!lookups.length) return;

    const methods = Object.entries(methodHash)
        .map(([key, value]) => {
            return {
                typeUrl: `/${context.ref.proto.package}.${value.requestType}`,
                TypeName: value.requestType,
                methodName: camel(key)
            }
        });

    // AMINO INTERFACES
    lookups.forEach(lookup => {
        const proto = lookup.obj;
        context.body.push(makeAminoTypeInterface({
            context: context.amino,
            proto,
            options: {
                aminoCasingFn: snake
            }
        }));
    });
    const protos = lookups.map(lookup => lookup.obj);

    // AMINO CONVERTER
    context.body.push(aminoConverter({
        name: 'AminoConverter',
        context: context.amino,
        root: context.ref.traversed,
        protos,
        options: {
            aminoCasingFn: snake
        }
    }))

    // add registry

    context.body.push(createTypeRegistry(methods));
    context.body.push(createRegistryLoader());

    // add methods
    context.body.push(
        toObjectWithPartialMethods(methods)
    );
    context.body.push(
        toObjectWithEncodedMethods(methods)
    );
    context.body.push(
        toObjectWithJsonMethods(methods)
    );
    context.body.push(
        toObjectWithToJSONMethods(methods)
    );
    context.body.push(
        toObjectWithFromJSONMethods(methods)
    );
    context.body.push(
        toObjectWithEncodedMethods(methods)
    );

    // add client

    // NEED BUNDLE
    // BUNDLE needs planning FILES/FILE structure

};

interface ParseRecur {
    context: TelescopeParseContext,
    obj: any;
    scope: string[];
    isNested: boolean;
}
export const parseRecur = ({
    context,
    obj,
    scope,
    isNested
}: ParseRecur) => {
    switch (obj.type) {
        case 'Type':
            return parseType(
                context, obj, scope, isNested
            );
        case 'Enum':
            return parseEnum(
                context, obj, scope, isNested
            );
        case 'Service':
            return parseService(
                context, obj, scope, isNested
            );
        case 'Field':
            console.log(obj);
            return;
        case 'Root':
        case 'Namespace':
            if (obj.nested) {
                return Object.keys(obj.nested).forEach(key => {
                    parseRecur({
                        context,
                        obj: obj.nested[key],
                        scope: [...scope, key],
                        isNested
                    });
                });
            } else {
                throw new Error('parseRecur() cannot find protobufjs Type')
            }
        default:
            console.log({ obj });
            throw new Error('parseRecur() cannot find protobufjs Type')
    }
};