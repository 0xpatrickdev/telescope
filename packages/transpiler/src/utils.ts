import { ProtoStore, ProtoRoot, ProtoRef, getObjectName } from '@osmonauts/proto-parser';
import { relative, dirname, extname } from 'path';

export const getRoot = (ref: ProtoRef): ProtoRoot => {
    if (ref.traversed) return ref.traversed;
    return ref.proto;
};

export const UTILS = {
    AminoHeight: '@osmonauts/helpers',
    AminoMsg: '@cosmjs/amino',
    AminoTypes: '@cosmjs/stargate',
    decodeBech32Pubkey: '@cosmjs/amino',
    defaultRegistryTypes: '@cosmjs/stargate',
    encodeBech32PubKey: '@cosmjs/amino',
    fromBase64: '@cosmjs/encoding',
    fromBech32: '@cosmjs/encoding',
    fromDuration: '@osmonauts/helpers',
    fromHex: '@cosmjs/encoding',
    fromJsonTimestamp: '@osmonauts/helpers',
    fromTimestamp: '@osmonauts/helpers',
    GeneratedType: '@cosmjs/proto-signing',
    isSet: '@osmonauts/helpers',
    Long: '@osmonauts/helpers', // exports Long and also calls the magic Long code
    // Long: {type: 'default', 'long'}
    OfflineSigner: '@cosmjs/proto-signing',
    omitDefault: '@osmonauts/helpers',
    Registry: '@cosmjs/proto-signing',
    SigningStargateClient: '@cosmjs/stargate',
    toBase64: '@cosmjs/encoding',
    toDuration: '@osmonauts/helpers',
    toTimestamp: '@osmonauts/helpers',
    bytesFromBase64: '@osmonauts/helpers',
    base64FromBytes: '@osmonauts/helpers',
    _m0: 'protobuf/minimal'
    // _m0: { type: 'namespace', path: 'protobuf/minimal' }
};


export const getRelativePath = (f1: string, f2: string) => {
    const rel = relative(dirname(f1), f2);
    let importPath = rel.replace(extname(rel), '');
    if (!/\//.test(importPath)) importPath = `./${importPath}`;
    return importPath;
}

export const insertUniq = (array: string[], el: string) => {
    if (!array.includes(el)) {
        array.push(el);
    }
}
