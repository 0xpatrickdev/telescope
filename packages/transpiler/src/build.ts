import { ProtoStore, ProtoRef } from '@osmonauts/proto-parser';
import {
    importStmt,
    AminoParseContext,
    createAminoConverter,
    createCreateProtoType,
    createObjectWithMethods,
    createProtoEnum,
    createProtoType,
    makeAminoTypeInterface,
    ProtoParseContext,
    // client
    createClient,
    GenericParseContext,
    // registry 
    createTypeRegistry,
    createRegistryLoader,
    // helper
    createHelperObject,
} from '@osmonauts/ast';
import { extname, relative, dirname } from 'path';
import { ImportHash, ServiceMutation, ServiceQuery } from './types';
import { UTILS, getRelativePath, insertUniq, getRoot } from './utils';

export const buildAllImports = (context: TelescopeParseContext, allImports?: ImportHash) => {

    if (!allImports) allImports = {};

    const utils = Object.keys({
        ...context.amino.utils,
        ...context.proto.utils
    });

    utils.forEach(util => {
        if (!UTILS.hasOwnProperty(util)) throw new Error('missing Util! ::' + util);
        allImports[UTILS[util]] = allImports[UTILS[util]] || [];
        insertUniq(allImports[UTILS[util]], util);
    });

    const parsedImports: ImportHash = context.amino.ref.traversed.parsedImports ?? {};

    Object.entries(parsedImports)
        .forEach(([filename, names]) => {
            const importPath = getRelativePath(context.ref.filename, filename);
            allImports[importPath] = allImports[importPath] || [];
            names.forEach(name => insertUniq(allImports[importPath], name));
        });

    const importStmts = Object.entries(allImports)
        .map(([importPath, names]) => {
            return importStmt(names, importPath);
        })

    return importStmts;
}

export const getMutations = (mutations: ServiceMutation[]) => {
    return mutations.map((mutation: ServiceMutation) => {
        return {
            typeUrl: `/${mutation.package}.${mutation.message}`,
            TypeName: mutation.message,
            methodName: mutation.methodName
        }
    });
};

export const getAminoProtos = (mutations: ServiceMutation[], store: ProtoStore) => {
    return mutations.map(mutation => {
        const ref = store.findProto(mutation.messageImport);
        return store.get(ref, mutation.message).obj;
    });
};

export const getAminoImports = (mutations: ServiceMutation[]) => {
    return mutations.map(mutation => {
        return {
            import: mutation.messageImport,
            name: mutation.message
        };
    });
};

interface BuildBaseTSClass {
    includeEncode: boolean;
    includeDecode: boolean;
    includeFromPartial: boolean;
    includeFromJSON: boolean;
    includeToJSON: boolean;
}

export interface TelescopeParseContext {
    proto: ProtoParseContext;
    amino: AminoParseContext;
    store: ProtoStore;
    ref: ProtoRef;
    parsedImports: Record<string, any>;
    body: any[];
    mutations: ServiceMutation[];
    queries: any[];
    types: any[];
    options: any;
}

export const buildBaseTypeScriptClass = (
    context: TelescopeParseContext,
    name: string,
    obj: any,
    {
        includeEncode,
        includeDecode,
        includeFromPartial,
        includeFromJSON,
        includeToJSON
    }: BuildBaseTSClass = {
            includeEncode: true,
            includeDecode: true,
            includeFromPartial: true,
            includeFromJSON: true,
            includeToJSON: true
        }) => {

    context.body.push(createProtoType(name, obj));
    context.body.push(createCreateProtoType(name, obj));
    context.body.push(createObjectWithMethods(context.proto, name, obj));
};

export const getAminoRelativeDeps = (mutations: ServiceMutation[], filename: string) => {
    return getAminoImports(mutations)
        .map(imp => {
            const f = filename;
            const f2 = imp.import;
            if (f === f2) return;
            const rel = relative(dirname(f), f2);
            let importPath = rel.replace(extname(rel), '');
            if (!/\//.test(importPath)) importPath = `./${importPath}`;
            return {
                ...imp,
                importPath
            };
        })
        .filter(Boolean)
        .reduce((m, v) => {
            m[v.importPath] = m[v.importPath] ?? [];
            if (!m[v.importPath].includes(v.name)) {
                m[v.importPath].push(v.name);
            }
            return m;
        }, {});
};
export class TelescopeParseContext implements TelescopeParseContext {
    constructor(ref: ProtoRef, store: ProtoStore) {
        this.proto = new ProtoParseContext();
        this.amino = new AminoParseContext(
            ref, store
        );
        this.ref = ref;
        this.store = store;
        this.parsedImports = {};
        this.body = [];
        this.queries = [];
        this.mutations = [];
        this.types = [];
    }

    hasMutations() {
        return this.mutations.length > 0;
    }

    addType(name: string, obj: any) {
        this.types.push({
            name,
            obj
        });
    }
    addMutation(mutation: ServiceMutation) {
        this.mutations.push(mutation);
    }
    addQuery(query: ServiceQuery) {
        this.queries.push(query);
    }

    // build main Class with methods
    buildBase() {
        this.types.forEach(typeReg => {
            const { name, obj } = typeReg;
            if (obj.type === 'Type') {
                buildBaseTypeScriptClass(this, name, obj);
            } else if (obj.type === 'Enum') {
                this.body.push(createProtoEnum(name, obj));
            } else {
                throw new Error('buildBase(): unknown type');
            }
        })
    }

    buildRegistry() {
        this.body.push(createTypeRegistry(getMutations(this.mutations)));
    }
    buildRegistryLoader() {
        this.body.push(createRegistryLoader());
    }
    buildAminoInterfaces() {
        const protos = getAminoProtos(this.mutations, this.store);
        protos.forEach(proto => {
            this.body.push(makeAminoTypeInterface({
                context: this.amino,
                proto
            }));
        });
    }
    buildAminoConverter() {
        this.body.push(createAminoConverter({
            name: 'AminoConverter',
            context: this.amino,
            root: this.ref.traversed,
            protos: getAminoProtos(this.mutations, this.store)
        }));
    }
    buildHelperObject() {
        // add methods
        this.body.push(createHelperObject({
            name: 'MessageComposer',
            mutations: getMutations(this.mutations)
        }));
    }
}