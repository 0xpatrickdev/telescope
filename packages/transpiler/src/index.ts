import * as t from '@babel/types';
import generate from '@babel/generator';
import { ProtoStore } from '@osmonauts/proto-parser';
import { buildAllImports, getServiceDependencies } from './imports';
import { TelescopeParseContext } from './build';
import { parse } from './parse';
import { bundlePackages } from './bundle';
import { writeFileSync } from 'fs';
import { join, dirname, resolve } from 'path';
import { sync as mkdirp } from 'mkdirp';
export interface TelescopeInput {
    protoDir: string;
    outPath: string;
}

export class TelescopeBuilder {
    store: ProtoStore;
    protoDir: string;
    outPath: string;

    contexts: TelescopeParseContext[] = [];

    constructor({ protoDir, outPath, store }: TelescopeInput & { store?: ProtoStore }) {
        this.protoDir = protoDir;
        this.outPath = outPath;

        this.store = store ?? new ProtoStore(protoDir);
        this.store.traverseAll();
    }

    context(ref) {
        const ctx = new TelescopeParseContext(
            ref, this.store
        );
        this.contexts.push(ctx);
        return ctx;
    }


    build(input: TelescopeInput) {

        // 0 REMEMBER!
        // YOU MUST SWITCH FROM in DIRECTORY to out DIRECTORY!!!
        // otherwise paths will break

        // ============



        // 1 get package bundle
        const bundled = bundlePackages(this.store, input);
        const packaged = bundled.reduce((m, bundle) => {
            const content = generate(t.program([
                ...bundle.importPaths,
                ...bundle.body
            ])).code


            const out = bundle.bundleFile;
            mkdirp(dirname(out));
            writeFileSync(out, content);

            // 2 search for all files that live in package
            console.log(`now looking for all [${bundle.base}]`);
            const baseProtos = this.store.getProtos().filter(ref => {
                return ref.filename.split('/')[0] === bundle.base
            });

            // 3 write out all TS files for package
            const packageContexts = baseProtos.map(ref => {
                const context = this.context(ref);
                parse(context);
                context.buildBase();

                // build BASE file
                const importStmts = buildAllImports(context);
                const prog = []
                    .concat(importStmts)
                    .concat(context.body);
                const gen = generate(t.program(prog));
                const filename = ref.filename.replace(/\.proto/, '.ts');
                const out = join(input.outPath, filename);
                mkdirp(dirname(out));
                writeFileSync(out, gen.code);

                return context;
            });

            // 4 find services w/mutations
            const serviceContexts = packageContexts.filter(context => context.mutations.length > 0);

            // 5 write out one amino helper for all contexts w/mutations

            const aminoHelpers = serviceContexts.map(c => {
                const filename = resolve(join(input.outPath, c.ref.filename)).replace(/\.proto$/, '.amino.ts'); ''
                // FRESH new context
                const ctx = new TelescopeParseContext(
                    c.ref,
                    c.store
                );

                // get mutations, services
                parse(ctx);

                // now let's amino...
                ctx.buildAminoInterfaces();
                ctx.buildAminoConverter();

                const serviceImports = getServiceDependencies(
                    ctx.mutations,
                    filename
                );

                // build file
                // DONT RENAME THE REF! you'll need to make a new one!
                // OR ELSE LATER the other build will use this name!
                // ctx.ref.filename = filename;
                const imports = buildAllImports(ctx, serviceImports);
                const content = generate(t.program([
                    ...imports,
                    ...ctx.body
                ])).code;
                mkdirp(dirname(filename));
                writeFileSync(filename, content);

                return filename;
            });

            // 6 write out one registry helper for all contexts w/mutations

            const registries = serviceContexts.map(c => {

                const filename = resolve(join(input.outPath, c.ref.filename)).replace(/\.proto$/, '.registry.ts'); ''
                // FRESH new context
                const ctx = new TelescopeParseContext(
                    c.ref,
                    c.store
                );

                // get mutations, services
                parse(ctx);

                ctx.buildRegistry();
                ctx.buildRegistryLoader();
                ctx.buildHelperObject();

                // SEE ABOVE - DONT RENAME THESE DIRECTLY
                // ctx.ref.filename = filename;
                const imports = buildAllImports(ctx);
                const content = generate(t.program([
                    ...imports,
                    ...ctx.body
                ])).code;
                mkdirp(dirname(filename));
                writeFileSync(filename, content);
                // console.log(filename);

                console.log(c.ref.filename);

                return filename;

            })

            // 7 write out one client for each base package, referencing the last two steps

            return m;
        }, {});

    }

    buildAll() {
        const hash = this.store.getProtos().map(ref => {
            const context = this.context(ref);
            parse(context);

            context.buildBase();

            // TODO don't loop through all protos...
            // loop through packgaes and put these into separate files
            if (context.mutations.length) {
                context.buildAminoInterfaces();
                context.buildAminoConverter();
                context.buildRegistry();
                context.buildRegistryLoader();
                context.buildHelperObject();
            }

            const importStmts = buildAllImports(context);

            const prog = []
                .concat(importStmts)
                .concat(context.body);

            const gen = generate(t.program(prog));

            const filename = ref.filename.replace(/\.proto/, '.ts');
            return {
                filename,
                content: gen.code
            };
        });
        return hash;
    }

    buildProto(path) {
        const ref = this.store.findProto(path);
        const context = this.context(ref);
        parse(context);
        context.buildBase();

        // generate(context, 'amino')
        // generate(context, 'proto')

        // maybe we can do something like this to separate out the context..
        // parse(context1, {
        //     amino: true
        // });

        // parse(context2, {


        //     registry: true
        // });

        // messages = c.toObjectWithPartialMethods(this.mutations);
        // registry = c.createTypeRegistry(this.mutations);
        // json = c.toObjectWithJsonMethods(this.mutations);
        // toJSON = c.toObjectWithToJSONMethods(this.mutations);
        // fromJSON = c.toObjectWithFromJSONMethods(this.mutations);
        // encoded = c.toObjectWithEncodedMethods(this.mutations);

        // this.outFiles = [];

        const importStmts = buildAllImports(context);

        const prog = []
            .concat(importStmts)
            .concat(context.body);

        const gen = generate(t.program(prog));
        return {
            gen,
            context
        }
    }
}

export default (input: TelescopeInput) => {
    const builder = new TelescopeBuilder(input);
    // const res = builder.buildProto('osmosis/gamm/v1beta1/tx.proto');
    // console.log(res);
};

