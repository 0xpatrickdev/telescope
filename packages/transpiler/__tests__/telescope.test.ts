import { TelescopeBuilder } from '../src';
import * as t from '@babel/types';
import { resolve, join, dirname, basename } from 'path';
import { ProtoStore, ProtoRef } from '@osmonauts/proto-parser'
import { bundleBaseRegistries, bundlePackages, bundleRegistries, getPackagesBundled, parseContextsForRegistry } from '../src/bundle'
import generate from '@babel/generator';
import { TelescopeParseContext } from '../src/build';
import { writeFileSync } from 'fs';
import { sync as mkdirp } from 'mkdirp';
import { buildAllImports, getServiceDependencies } from '../src/imports';
import { expectCode } from '../test-utils/index'
import { parse } from '../src/parse';

const store = new ProtoStore(__dirname + '/../../../__fixtures__/chain1');
store.traverseAll();

const input = {
    outPath: __dirname + '/../../../__fixtures__/output1',
    protoDir: __dirname + '/../../../__fixtures__/chain1'
};
const telescope = new TelescopeBuilder(input);

describe('bundle package registries and root file names', () => {
    // console.log(telescope);
    const hash = telescope.buildAll();
    const outPath = resolve(input.outPath);
    hash.forEach(({ filename, content }) => {
        const out = join(outPath, filename);
        // console.log(out);
        mkdirp(dirname(out));
        writeFileSync(out, content);
    })
    it('bundleRegistries', () => {
        const registries = bundleRegistries(telescope);
        const result = registries.map(reg => ({
            ['package']: reg.package,
            contexts: parseContextsForRegistry(reg.contexts)
        }))
        // console.log(JSON.stringify(result, null, 2));
    })

    it('bundleBaseRegistries', () => {
        const registries = bundleBaseRegistries(telescope);
        const result = registries.map(reg => ({
            base: reg.base,
            pkgs: reg.pkgs.map(
                obj => {
                    return {
                        ['package']: obj.package,
                        contexts: parseContextsForRegistry(obj.contexts)
                    }
                }
            )
        }));
        // console.log(JSON.stringify(result, null, 2));

    })

    it('bundle deps from a context', () => {
        const context = telescope.contexts.find(context => context.ref.filename === 'osmosis/gamm/v1beta1/tx.proto');
        const imports2 = getServiceDependencies(
            context.mutations,
            join(dirname(context.ref.filename), 'bundle.ts')
        );
        expect(imports2).toMatchSnapshot();
        console.log(imports2);

        const evenMoreImports = buildAllImports(context, imports2);
        console.log(generate(t.program(evenMoreImports)).code)
        expectCode(t.program(evenMoreImports))
    });
    it.only('bundle deps by creating a context', () => {

        const context = telescope.contexts.find(context => context.ref.filename === 'osmosis/gamm/v1beta1/tx.proto');

        // FRESH new context
        const ctx = new TelescopeParseContext(
            context.ref,
            context.store
        );

        expect(ctx.mutations.length).toBe(0);

        // get mutations, services
        parse(ctx);

        // now we have some mutations
        expect(ctx.mutations.length).toBe(8);

        // but still no imports...
        expect(ctx.amino.imports.length).toBe(0)
        expect(ctx.proto.imports.length).toBe(0)

        // now let's amino...
        ctx.buildAminoInterfaces();
        ctx.buildAminoConverter();

        // now we have amino imports
        expect(ctx.amino.imports.length).toBe(18);
        expect(ctx.proto.imports.length).toBe(0);

        console.log(ctx.amino.imports);


        const serviceImports = getServiceDependencies(
            context.mutations,
            join(dirname(context.ref.filename), 'telescope/amino-converter.ts')
        );
        expect(serviceImports).toMatchSnapshot();
        console.log(serviceImports);

        // MAKE A FAKE REF

        const ref: ProtoRef = {
            absolute: 'some/path',
            filename: 'osmosis/gamm/v1beta1/telescope/amino-converter.proto',
            proto: {},
        };
        // context.ref.filename = 'osmosis/gamm/v1beta1/amino-converter.proto';
        context.ref = ref;
        const evenMoreImports = buildAllImports(context, serviceImports);
        console.log(generate(t.program(evenMoreImports)).code)
        expectCode(t.program(evenMoreImports))

        // make AST
        const ast = t.program([...evenMoreImports, ...ctx.body])

        console.log(generate(ast).code);

    });
})





