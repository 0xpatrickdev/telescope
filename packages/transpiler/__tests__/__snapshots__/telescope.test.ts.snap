// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bundle package registries and root file names bundle deps by creating a context 1`] = `
Object {
  "../tx": Array [
    "MsgJoinPool",
    "MsgExitPool",
    "MsgSwapExactAmountIn",
    "MsgSwapExactAmountOut",
    "MsgJoinSwapExternAmountIn",
    "MsgJoinSwapShareAmountOut",
    "MsgExitSwapExternAmountOut",
    "MsgExitSwapShareAmountIn",
  ],
}
`;

exports[`bundle package registries and root file names bundle deps by creating a context 2`] = `
"import { Coin } from \\"../../../../cosmos/base/v1beta1/coin\\";
import { AminoMsg } from \\"@cosmjs/amino\\";
import { Long, isSet, Exact, DeepPartial } from \\"@osmonauts/helpers\\";
import * as _m0 from \\"protobuf/minimal\\";
import { MsgJoinPool, MsgExitPool, MsgSwapExactAmountIn, MsgSwapExactAmountOut, MsgJoinSwapExternAmountIn, MsgJoinSwapShareAmountOut, MsgExitSwapExternAmountOut, MsgExitSwapShareAmountIn } from \\"../tx\\";"
`;
